### ISSUE TITLE
AST Translator Assertion failed

### ISSUE THREAD

**Issue by olligobber** [souffle-lang/souffle#800]
When compiling the code [dfa.souffle](https://github.com/souffle-lang/souffle/files/2627295/dfa.souffle.txt) with a fully tested unmodified version of souffle, the following error occurs:
```
souffle: AstTranslator.cpp:240: virtual std::unique_ptr<souffle::RamValue> souffle::AstTranslator::translateValue(const souffle::AstArgument*, const souffle::AstTranslator::ValueIndex&)::ValueTranslator::visitVariable(const souffle::AstVariable&): Assertion `index.isDefined(var) && "variable not grounded"' failed.
Aborted (core dumped)
```

**Comment by brodyf** [comment]
More minimal example:

```
.decl Z(a: symbol)
.decl R(a: number)
.output R(IO=stdout)

Z("a").

R(0) :- Z(z), l=strlen(z), z=substr("a",0,l).
```

**Comment by azreika** [comment]
The program works if you reorder the constraints:

```
R(0) :- Z(z), z=substr("a",0,l), l = strlen(z).
```

because then `ResolveAliasesTransformer` rewrites this to:

```
R(0) :- Z(z), z=substr("a",0,strlen(z)).
```

Here's a minimal example that doesn't involve string functors:

```
.decl a(x:number)
a(0).

.decl q(x:number) output
q(0) :- a(x), y = 1*x, x = 1*y.
```

Again, reordering the constraints gets rid of the problem, because

```
q(0) :- a(x), x = 1*y, y = 1*x.
```

is transformed into:

```
q(0) :- a(x), x = 1*(1*x).
```

Should look into why these variables aren't being resolved properly in the `ResolveAliasesTransformer` in some orders.

**Comment by b-scholz** [comment]
Semantically, this example is interesting because it has an ungrounded variable l. 

However, we have a substitution transformation, which may or may not fire depending on context. An illegal rule (i.e. a rule with ungrounded variables) becomes legal after substitution.

**Comment by azreika** [comment]
Looking at the rule:

```
q(0) :- a(x), y = 1*x, x = 1*y.
```

Although `y` is ungrounded, the groundedness analysis (by the semantic checker) lets this pass: the value of `y` can be fixed by the constraint `y = 1*x`, since `x` is grounded. The problem during alias resolution is that the constraint `x = 1*y` is resolved first, however. The result is that all appearances of `x` are replaced with `1*y`, giving us the rule:

```
q(0) :- a(1*y), y = 1*(1*y), 1*y = 1*y.
```

Now, `y` isn't grounded, since it never appears functorless in a non-negated body atom, so this breaks during translation.

Assuming the groundedness analysis is correct, this can be fixed by not trying to resolve aliases of the form `x = ...` where `x` already appears in a grounding atom (e.g. `a(x)`). That way, since all other introduced variables do not violate the groundedness analysis, resolving the remaining equality constraints in any order should eventually get rid of the ungrounded variables (like `y` in the above example).

In this case, the produced rule is instead:

```
q(0) :- a(x), 1*x = 1*x, x = 1*(1*x).
```
which is valid.

I'll pull up a fix soon.

**Comment by azreika** [comment]
Fixed by PR #806.
